src/java.base/share/classes/java/util/stream/Gatherer.java

```diff
@@ -0,0 +1,566 @@
+         *          never return false again for the same instance.
+         *
+         * By default this method returns {@code false}.
+         *
+         * @return {@code true} if this Downstream is known not to want any
+         *         more elements sent to it, {@code false} if otherwise
+         */
+        default boolean isRejecting() { return false; }
+    }
+
+    /**
+     * An Integrator receives elements and processes them,
+     * optionally using the supplied state, and optionally sends incremental
+     * results downstream.
+     *
+     * @param <A> the type of state used by this integrator
+     * @param <T> the type of elements this integrator consumes
+     * @param <R> the type of results this integrator can produce
+     */
+    @FunctionalInterface
+    @PreviewFeature(feature = PreviewFeature.Feature.STREAM_GATHERERS)
+    interface Integrator<A, T, R> {
+        /** Integrate is the method which given:
+         * the current state, the next element, and a downstream object;
+         * performs the main logic -- potentially inspecting and/or updating
+         * the state, optionally sending any number of elements downstream
+         * -- and then returns whether more elements are to be consumed or not.
+         *
+         * @param state The state to integrate into
+         * @param element The element to integrate
+         * @param downstream The downstream object of this integration,
+         *                   returns false if doesn't want any more elements
+         * @return {@code true} if subsequent integration is desired,
+         *         {@code false} if not
+         */
+        boolean integrate(A state, T element, Downstream<? super R> downstream);
+
+        /**
+         * Factory method for turning Integrator-shaped lambdas into
+         * Integrators.
+         *
+         * @param integrator a lambda to create as Integrator
+         * @return the given lambda as an Integrator
+         * @param <A> the type of elements this integrator receives
+         * @param <T> the type of initializer this integrator consumes
+         * @param <R> the type of results this integrator can produce
+         */
+        @ForceInline
+        static <A, T, R> Integrator<A, T, R> of(Integrator<A, T, R> integrator) {
+            return integrator;
+        }
+
+        /**
+         * Factory method for turning Integrator-shaped lambdas into
+         * {@link Greedy} Integrators.
+         *
+         * @param greedy a lambda to create as Integrator.Greedy
+         * @return the given lambda as a Greedy Integrator
+         * @param <A> the type of elements this integrator receives
+         * @param <T> the type of initializer this integrator consumes
+         * @param <R> the type of results this integrator can produce
+         */
+        @ForceInline
+        static <A, T, R> Greedy<A, T, R> ofGreedy(Greedy<A, T, R> greedy) {
+            return greedy;
+        }
+
```

Comment:
Since [GH‑16213] was closed in favour of adding diamond inference to casts (e.g.: `(Integrator<>) Foo::bar` or `(Integrator.Greedy<>) Foo::bar`), these should probably be removed as well:
```suggestion
```

[GH‑16213]: https://github.com/openjdk/jdk/pull/16213

Reply:
@ExE-Boss Since there is no diamond casts yet, and this is a Preview, I think it is safe enough to keep things as is and re-assess if/when such capabilities are supported.